hugr-core/src/extension/op_def.rs

```diff
@@ -142,26 +139,80 @@ impl CustomValidator {
     ) -> Self {
         Self {
             poly_func: poly_func.into(),
-            validate: Box::new(validate),
+            validate: Some(Box::new(validate)),
         }
     }
 }
 
 /// The two ways in which an OpDef may compute the Signature of each operation node.
-#[derive(serde::Deserialize, serde::Serialize)]
 pub enum SignatureFunc {
     // Note: except for serialization, we could have type schemes just implement the same
     // CustomSignatureFunc trait too, and replace this enum with Box<dyn CustomSignatureFunc>.
     // However instead we treat all CustomFunc's as non-serializable.
     /// A PolyFuncType (polymorphic function type), with optional custom
     /// validation for provided type arguments,
-    #[serde(rename = "signature")]
     PolyFuncType(CustomValidator),
-    #[serde(skip)]
     /// A custom binary which computes a polymorphic function type given values
     /// for its static type parameters.
     CustomFunc(Box<dyn CustomSignatureFunc>),
+    /// Declaration specified a custom binary but it was not provided.
+    MissingComputeFunc,
```

Comment:
What about a missing custom validation func? In either case I think we just have to trust the cache, maybe with a warning, so I guess if the warning is the same then we don't have to distinguish, is that the plan?

Reply:
I'm not sure - in the case of missing validation function do you trust the cached signature or use the type scheme to generate the signature and check against cache as you would without custom validation?

Reply:
Good point. Given the validation function can only say one of two things - invalid, or use the type scheme - you can try the latter (which might reject if TypeArgs don't match the TypeParams, say) and see if that matches; that might get you an error, but even if the typescheme says ok, if there's a binary validation function that you haven't got, then that still has to be a warning